### PR TITLE
[EXPLORER] System tray fixes when using large taskbar icons

### PR DESCRIPTION
**Pull request by cbialorucki**
## Purpose
- Provides Windows 7 style system tray icon spacing when using large taskbar icons.
- Fixes the clock spacing when the day of the week is shorter than the date.

Before:
Classic Theme Before

After:
Classic Theme After
Themed After (Luna used, pictured on ReactOS)

A comparison against the system tray in Windows 7. It's not pixel-perfect, but very close.
image

Jira issue: [CORE-11698](https://jira.reactos.org/browse/CORE-11698)

## Proposed changes
- Increase padding around system tray icons when using large taskbar icons.
- When using large taskbar icons, push the clock text further to the right, which increases the spacing to the left between the clock and notification icons. This matches the behavior of Windows 7. When small taskbar icons are used this behavior is ignored.
- When ```bPerferDateOverWeekday``` is enabled, check the clock spacing against the date when only two lines are visible.

## TODO (in separate PRs)
- Add the option to hide notification icons.
- Figure out a better way to space the notification icons when more than four lines are visible.

### CODE REVIEW

File: base/shell/explorer/syspager.cpp [hunk 1]
```diff
@@ -1256,6 +1256,11 @@ void CNotifyToolbar::Initialize(HWND hWndParent, CBalloonQueue * queue)
     tbm.cyBarPad = 1;
     tbm.cxButtonSpacing = 1;
     tbm.cyButtonSpacing = 1;
+    if (!g_TaskbarSettings.bSmallIcons)
+    {
+        tbm.cxPad = GetSystemMetrics(SM_CYSMICON) * 0.5;
```
Comment (learn-more) [inline base/shell/explorer/syspager.cpp:1261]: This should probably be `SM_CXSMICON`

Comment (learn-more) [reply base/shell/explorer/syspager.cpp:1261]: Where do these magic constants come from?

Comment (HBelusca) [inline base/shell/explorer/syspager.cpp:1261]:
```suggestion
        tbm.cxPad = GetSystemMetrics(SM_CXSMICON) / 2;
```

File: base/shell/explorer/syspager.cpp [hunk 2]
```diff
@@ -1256,6 +1256,11 @@ void CNotifyToolbar::Initialize(HWND hWndParent, CBalloonQueue * queue)
...
+        tbm.cyPad = GetSystemMetrics(SM_CYSMICON) * 0.5;
```
Comment (HBelusca) [inline base/shell/explorer/syspager.cpp:1262]:
```suggestion
        tbm.cyPad = GetSystemMetrics(SM_CYSMICON) / 2;
```

File: base/shell/explorer/syspager.cpp [hunk 3]
```diff
@@ -1398,10 +1403,18 @@ void CSysPagerWnd::GetSize(IN BOOL IsHorizontal, IN PSIZE size)
     INT cyButton = GetSystemMetrics(SM_CYSMICON) + 2;
     INT cxButton = GetSystemMetrics(SM_CXSMICON) + 2;
     int VisibleButtonCount = Toolbar.GetVisibleButtonCount();
+    if (!g_TaskbarSettings.bSmallIcons)
+    {
+        cyButton = GetSystemMetrics(SM_CYSMICON) * 1.5;
+        cxButton = GetSystemMetrics(SM_CXSMICON) * 1.5;
```
Comment (HBelusca) [inline base/shell/explorer/syspager.cpp:1409]:
```suggestion
        cyButton = 3 * GetSystemMetrics(SM_CYSMICON) / 2;
        cxButton = 3 * GetSystemMetrics(SM_CXSMICON) / 2;
```
or
```suggestion
        cyButton = MulDiv(GetSystemMetrics(SM_CYSMICON), 3, 2);
        cxButton = MulDiv(GetSystemMetrics(SM_CXSMICON), 3, 2);
```

File: base/shell/explorer/syspager.cpp [hunk 4]
```diff
@@ -1398,10 +1403,18 @@ void CSysPagerWnd::GetSize(IN BOOL IsHorizontal, IN PSIZE size)
...
+    }
 
     if (IsHorizontal)
     {
-        rows = max(size->cy / cyButton, 1);
+        if (g_TaskbarSettings.bSmallIcons)
+            rows = max(size->cy / cyButton, 1);
+        else
+            rows = max(size->cy / (cyButton * 1.5), 1);
```
Comment (HBelusca) [inline base/shell/explorer/syspager.cpp:1417]:
```suggestion
            rows = max(size->cy / (3 * cyButton / 2), 1);
```

File: base/shell/explorer/trayclock.cpp [hunk 5]
```diff
@@ -267,6 +267,10 @@ WORD CTrayClockWnd::GetMinimumSize(IN BOOL Horizontal, IN OUT PSIZE pSize)
     if (!LinesMeasured)
         return 0;
 
+    /* Prevents the date from being cut off when the day of the week is shorter than the date. */
```
Comment (HBelusca) [inline base/shell/explorer/trayclock.cpp:270]:
```suggestion
    /* Prevents the date from being cut off when the day of the week is shorter than the date */
```
By "date" you mean the "dd/mm/yyyy" thing (first line of the clock in the screenshots)?

**Review by HBelusca — COMMENTED** [review]
It's interesting to think of the first screenshot ("Before") as a hypothetical "compact system tray icons" setting :thinking:

File: base/shell/explorer/syspager.cpp [hunk 6]
```diff
@@ -1256,6 +1256,11 @@ void CNotifyToolbar::Initialize(HWND hWndParent, CBalloonQueue * queue)
     tbm.cyBarPad = 1;
     tbm.cxButtonSpacing = 1;
     tbm.cyButtonSpacing = 1;
+    if (!g_TaskbarSettings.bSmallIcons)
```
Comment (HBelusca) [inline base/shell/explorer/syspager.cpp:1259]: Might be moved just below the original cx/yPad setting done few lines above (1253)

File: base/shell/explorer/syspager.cpp [hunk 7]
```diff
@@ -1398,10 +1403,18 @@ void CSysPagerWnd::GetSize(IN BOOL IsHorizontal, IN PSIZE size)
     INT cyButton = GetSystemMetrics(SM_CYSMICON) + 2;
     INT cxButton = GetSystemMetrics(SM_CXSMICON) + 2;
     int VisibleButtonCount = Toolbar.GetVisibleButtonCount();
+    if (!g_TaskbarSettings.bSmallIcons)
```
Comment (HBelusca) [inline base/shell/explorer/syspager.cpp:1406]: Might be moved before initialization of VisibleButtonCount.

File: base/shell/explorer/trayntfy.cpp [hunk 8]
```diff
@@ -213,7 +213,7 @@ class CTrayNotifyWnd :
 
             ::SetWindowPos(m_hwndClock,
                 NULL,
-                ptClock.x,
+                g_TaskbarSettings.bSmallIcons ? ptClock.x : ptClock.x + 5,
```
Comment (HBelusca) [inline base/shell/explorer/trayntfy.cpp:216]: Regarding the "5" , could you check whether, on Windows, the traybar icons stay separated the same amount even if you artificially increase the SM_CXBORDER value (window border, via the display cpl) ?

Comment (HBelusca) [reply base/shell/explorer/trayntfy.cpp]: Thanks, so using a static value seems to make sense.

**Review by HBelusca — APPROVED** [review]

**Comment by cbialorucki** [comment]
> It's interesting to think of the first screenshot ("Before") as a hypothetical "compact system tray icons" setting 🤔

Should we implement a `bCompactTrayIcons` registry tweak? Thumbs up for yes, thumbs down for no.

**Comment by cbialorucki** [comment]
I added a registry setting for compact tray icons (Looks the same as the "before" picture when enabled.) This is a DWORD value in `HKCU\SOFTWARE\Microsoft\Windows\CurrentVersion\Explorer\Advanced\CompactTrayIcons`.

File: base/shell/explorer/trayclock.cpp [hunk 9]
```diff
@@ -571,10 +571,11 @@ VOID CTrayClockWnd::PaintLine(IN HDC hDC, IN OUT RECT *rcClient, IN UINT LineNum
     if (LineSizes[LineNumber].cx == 0)
         return;
 
-    INT HorizontalShift = g_TaskbarSettings.bSmallIcons ? 0 : TRAY_CLOCK_WND_SPACING_X;
+    INT HShift = g_TaskbarSettings.bSmallIcons ||
+                 g_TaskbarSettings.bCompactTrayIcons ? 0 : TRAY_CLOCK_WND_SPACING_X;
```
Comment (HBelusca) [inline base/shell/explorer/trayclock.cpp:575]:
```suggestion
    INT HShift = ((g_TaskbarSettings.bSmallIcons ||
                   g_TaskbarSettings.bCompactTrayIcons) ? 0 : TRAY_CLOCK_WND_SPACING_X);

**Review by HBelusca — APPROVED** [review]

**Comment by JoachimHenze** [comment]
Is there also an option to have taskbar with 2 rows for the running apps and then combining that with single-rowed-compact-mode for the systray? I guess not yet, but it might be an interesting thing to have at some point.

At least we seem to be able to have 2 rows for the running apps, and that combined with 2-rows-compact-mode for the systray. I never liked how Win7 wastes the space in the taskbar.

**Comment by cbialorucki** [comment]
> Is there also an option to have taskbar with 2 rows for the running apps and then combining that with single-rowed-compact-mode for the systray? I guess not yet, but it might be an interesting thing to have at some point.

I reworked the logic so that it is now possible to set compact tray icons on or off independent of the small taskbar icons setting. By default, if the registry value is missing it will be compact when using small icons and not compact when using large icons.

**Review by HBelusca — APPROVED** [review]

**Review by binarymaster — APPROVED** [review]

**Pull request merged**